sdk/iot/pnp/inc/az_iot_pnp_client.h

```diff
@@ -0,0 +1,334 @@
+ *                     reasons we recommend using one key per device instead of using a global
+ *                     policy key.
+ * @param[in] mqtt_password An empty #az_span with sufficient capacity to hold the MQTT password.
+ * @param[out] out_mqtt_password The output #az_span containing the MQTT password.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_pnp_client_sas_password_get(
+    az_iot_pnp_client const* client,
+    az_span base64_hmac_sha256_signature,
+    az_span key_name,
+    az_span mqtt_password,
+    az_span* out_mqtt_password);
+
+/*
+ *
+ * PnP Telemetry APIs
+ *
+ */
+
+/**
+ * @brief Azure IoT PnP telemetry options.
+ *
+ */
+typedef struct az_iot_pnp_client_telemetry_options
+{
+  struct
+  {
+    // OPEN: I can't find samples in Azure SDK modeling this.  Closest is
+    // sdk\storage\blobs\inc\az_storage_blobs.h where it has 'az_span option" in
+    // az_storage_blobs_blob_download_options that's ignored in practice.
+    int _not_used;
+  } _internal;
+} az_iot_pnp_client_telemetry_options;
+
+/**
+ * @brief Gets the default Azure IoT PnP telemetry options.
+ * @details Call this to obtain an initialized #az_iot_pnp_client_telemetry_options structure that
+ * can be afterwards modified and passed to #az_iot_pnp_client_telemetry_publish_topic_get.
+ *
+ * @return #az_iot_pnp_client_telemetry_options.
+ */
+AZ_NODISCARD az_iot_pnp_client_telemetry_options az_iot_pnp_client_telemetry_options_default();
+
+/**
+ * @brief Gets the MQTT topic that must be used for device to cloud telemetry messages.
+ * @note Telemetry MQTT Publish messages must have QoS At Least Once (1).
+ * @note This topic can also be used to set the MQTT Will message in the Connect message.
+ *
+ * Should the user want a null terminated topic string, they may allocate a buffer large enough
+ * to fit the topic plus a null terminator. They must set the last byte themselves or zero
+ * initialize the buffer.
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[in] component_name An #az_span specifying the component name to publish telemetry on.
+ * @param[in] content_type An #az_span specifying the content type of the message payload.
+                           Only AZ_SPAN_FROM_STR("application/json") is currently supported.
+ * @param[in] encoding_type An #az_span specifying the encoding type of the message payload.
+                           Only AZ_SPAN_FROM_STR("utf-8") is currently supported.
+ * @param[in] options A reference to an #az_iot_pnp_client_telemetry_options structure. Can be NULL.
+ * @param[in] mqtt_topic An empty #az_span with sufficient capacity to hold the MQTT topic.
+ * @param[out] out_mqtt_topic The output #az_span containing the MQTT topic.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_pnp_client_telemetry_publish_topic_get(
+    az_iot_pnp_client const* client,
+    az_span component_name,
+    az_span content_type,
```

Comment:
;tl-dr - I'm going to do what you suggest

Details:
I *do* think these should be optional and NOT required.  New thread started offline to make sure there's no surprises there but pushing this as plan.

Here's what I'm thinking generally:
* Move to options to the `client` since overwhelming majority if not all of time it'll be set.  And certainly for GA that will be the case RE only one type.
* If(!) it turns out convenience layer needs to mix/match JSON and some other binding -- and there's no clear scenario for that yet! - then we'd extend out the telemetry_options field to let it over-ride the "global".  So we're not architected into a corner here, but are designing for common case.